File: God Zuo/class062 bfs/My05.py
# 测试链接 : https://leetcode.cn/problems/trapping-rain-water-ii/\
import heapq
class Solution:
    def trapRainWater(self, heightMap: list[list[int]]) -> int:
        if len(heightMap) < 3 or len(heightMap[0]) < 3:
            return 0
        m,n=len(heightMap),len(heightMap[0])
        visited=[[False]*n for _ in range(m)]
        heap=[]
        for i in range(m):
            for j in range(n):
                if i==0 or i==m-1 or j==0 or j==n-1:
                    heapq.heappush(heap,(heightMap[i][j],i,j))
                    visited[i][j]=True
        ans=0
        move=[-1,0,1,0,-1]
        while heap:
            w,r,c=heapq.heappop(heap)
            ans+=w-heightMap[r][c]
            for i in range(4):
                tr=r+move[i]
                tc=c+move[i+1]
                if 0<=tr<m and 0<=tc<n and not visited[tr][tc]:
                    heapq.heappush(heap,((w if w >heightMap[tr][tc] else heightMap[tr][tc]),tr,tc))
                    visited[tr][tc]=True
        return ans
    

File: God Zuo/class069 三维动态规划/My05.cpp
// 扰乱字符串
// 使用下面描述的算法可以扰乱字符串 s 得到字符串 t ：
// 步骤1 : 如果字符串的长度为 1 ，算法停止
// 步骤2 : 如果字符串的长度 > 1 ，执行下述步骤：
//        在一个随机下标处将字符串分割成两个非空的子字符串
//        已知字符串s，则可以将其分成两个子字符串x和y且满足s=x+y
//        可以决定是要 交换两个子字符串 还是要 保持这两个子字符串的顺序不变
//        即s可能是 s = x + y 或者 s = y + x
//        在x和y这两个子字符串上继续从步骤1开始递归执行此算法
// 给你两个 长度相等 的字符串 s1 和 s2，判断 s2 是否是 s1 的扰乱字符串
// 如果是，返回true ；否则，返回false
// 测试链接 : https://leetcode.cn/problems/scramble-string/

#include<bits/stdc++.h>
using namespace std;

class Solution{
public:
    int dp[35][35][35] = {};
    bool f(const string& s1,const string& s2 ,int l1,int l2,int len){
        if(len==1)
            return s1[l1] == s2[l2];
        if(dp[l1][l2][len]!=0)
            return dp[l1][l2][len] == 1;
        bool ans = false;
        for (int k = 1; k < len;++k){
            if(f(s1,s2,l1,l2,k) && f(s1,s2,l1+k,l2+k,len-k)){
                ans = true;
                break;
            }
        }
        if(!ans){
            for (int k = 1;k<len;++k){
                if (f(s1, s2, l1, l2 +len- k, k) && f(s1, s2, l1 + k, l2, len - k)) {
                    ans = true;
                    break;
                }
            }
        }
        dp[l1][l2][len] = ans ? 1 : -1;
        return ans;
    }

    bool isScramble(string s1, string s2){
        if(s1.length()!=s2.length())
            return false;
        int n = s1.length();
        for (int i = 0; i < n;++i){
            for (int j = 0; j < n;++j){
                for (int k = 0; k <= n;++k)
                    dp[i][j][k] = 0;
            }
        }
        return f(s1, s2, 0, 0, n);
    }
};